linchpin/cli/__init__.py

```diff
@@ -173,6 +174,99 @@ def _get_data_path(self):
         return data_w_path
 
 
+    def _write_distilled_context(self, run_data):
+        """
+        This method takes all of the provided run_data, loops through the
+        distiller section of the linchpin.constants and writes out the
+        linchpin.context (name TBD) to the 'resources' directory.
+
+        """
+
+        dist_roles = self.get_cfg('distiller')
+
+        resources_path = self.get_evar('resources_folder')
+        context_path = '{0}/{1}'.format(self.workspace, resources_path)
+        if not os.path.exists(context_path):
+            os.makedirs(context_path)
+        context_file = '{0}/{1}'.format(context_path, 'linchpin.distilled')
+
+        roles = []
+        dist_data = {}
+
+
+        # get roles used
+        for target, data in run_data.iteritems():
+            inputs = data.get('inputs')
+            topo = inputs.get('topology_data')
+            res_grps = topo.get('resource_groups')
+
+            for group in res_grps:
+                res_defs = group.get('resource_definitions')
+                for rd in res_defs:
+                    roles.append(rd.get('role'))
+
+            fields = {}
+            outputs = data.get('outputs')
+            resources = outputs.get('resources')
+
+            for dist_role, flds in dist_roles.iteritems():
+                if dist_role in roles:
+                    for f in flds.split(','):
+                        if '.' not in f:
+                            fld = fields.get('single', [])
+                            fields[f] = None
+                        else:
+                            k, v = f.split('.')
+                            fld = fields.get(k, [])
+                            if ':' in v:
+                                subfields = {}
+                                k2, val = v.split(':')
+                                subfld = subfields.get(k2, [])
+                                subfld.append(val)
+                                subfields[k2] = subfld
+                                fld.append(subfields)
+                                fields[k] = fld
+                            else:
+                                fld.append(v)
+                                fields[k] = fld
+
+            try:
+                for res in resources:
+                    for k, v in fields.iteritems():
+                        res_data = {}
+                        if not v:
+                            dd = dist_data.get(target, {})
+                            if len(dd):
+                                for items in dd:
+                                    if k in items.keys():
+                                        items[k].extend(res.get(k))
```

Comment:
Can you explain what this loop is doing?  Seems a little odd to me to be modifying the dictionary you are iterating over.

Reply:
Sure. There are actually two dictionaries here. The first pulls out the fields which need to be grabbed from the particular provider. The second matches those items and  puts them into the dictionary.

Reply:
ok - I get it now.